### PR TITLE
Extend XLA tensor view to support partial/sliced updates.

### PR DESCRIPTION
**Pull request by dlibenzi**

### CODE REVIEW

File: torch_xla/csrc/data_ops.cpp [hunk 1]
```diff
@@ -291,4 +292,31 @@ std::vector<xla::XlaOp> BuildSplit(
...
+      XlaHelpers::NormalizeShape(source, input_shape.rank());
+  std::vector<xla::XlaOp> start_indices;
+  for (auto index : base_indices) {
+    start_indices.push_back(
+        XlaHelpers::ScalarValue<xla::int64>(index, input.builder()));
+  }
+  return xla::DynamicUpdateSlice(input, reshaped_source, start_indices);
+}
+
+xla::XlaOp BuildGenericSlice(
```
Comment (asuhan) [inline torch_xla/csrc/data_ops.cpp]: I don't like this `Generic` cop-out name. Could this be `BuildReadSlice`? And maybe rename `BuildUpdateSlice` to `BuildWriteSlice` for better name duality.

File: torch_xla/csrc/helpers.h [hunk 2]
```diff
@@ -140,6 +140,14 @@ class XlaHelpers {
 
   static xla::XlaComputation CreateMaxComputation(xla::PrimitiveType type);
 
+  // Returns an XLA operation which is a reshape to the expected rank, by
+  // appending 1s to the major dimension. If offset is greater than zero, 1s
+  // will be prepened to the minor dimension as well.
+  // Expected condition: rank(input) + offset <= expected_rank
+  static xla::XlaOp NormalizeShape(const xla::XlaOp& input,
```
Comment (asuhan) [inline torch_xla/csrc/helpers.h]: I think `ReshapeToRank` would be a better name.

File: torch_xla/csrc/view.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "tensorflow/compiler/xla/xla_client/util.h"
+#include "torch_xla/csrc/ops/generic_slice.h"
+#include "torch_xla/csrc/ops/update_slice.h"
+#include "torch_xla/csrc/ops/view.h"
+
+namespace torch_xla {
+namespace {
+
+bool IsNarrow(const ViewInfo& view_info) {
```
Comment (asuhan) [inline torch_xla/csrc/view.cpp]: I'm not sure I like the fact that the only way to distinguish between what's two distinct operators (view and narrow) is so implicit. Could we store the kind symbol in the `ViewInfo` as well?

File: torch_xla/csrc/view.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+void Alias::Update(ir::Value ir_value, std::vector<ViewInfo> view_infos) {
+  updates_.push_back({std::move(ir_value), std::move(view_infos)});
+}
+
+ir::Value Alias::CreateUpdateChain() const {
+  ir::Value result = ir_value_;
+  for (auto& update_data : updates_) {
+    result = ApplyUpdate(result, update_data);
+  }
+  return result;
```
Comment (asuhan) [inline torch_xla/csrc/view.cpp]: Are we sure we never need to clear `updates_`? Maybe we should have a comment with the explanation why.

File: torch_xla/csrc/data_ops.cpp (hunk 1 above)
Comment (dlibenzi) [reply torch_xla/csrc/data_ops.cpp]: Renamed BuildGenericSlice to Build Slice.

File: torch_xla/csrc/helpers.h (hunk 2 above)
Comment (dlibenzi) [reply torch_xla/csrc/helpers.h]: SG

File: torch_xla/csrc/view.cpp (hunk 3 above)
Comment (dlibenzi) [reply torch_xla/csrc/view.cpp]: So far we only support narrow and view/reshape, and this is what drives the two different update policies (require dyn-update-slice vs reshape).

File: torch_xla/csrc/view.cpp (hunk 4 above)
Comment (dlibenzi) [reply torch_xla/csrc/view.cpp]: It is OK not clear clear, but even though the update chain might still contain a long graph, I think we can save some time in not rebuilding it.
So yes, we can purge the updates and cache the IR value.

**Pull request merged**